src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergWrites.cpp

```diff
@@ -0,0 +1,772 @@
+    , object_storage(object_storage_)
+    , context(context_)
+    , configuration(configuration_)
+    , format_settings(format_settings_)
+    , filename_generator(configuration_->getPath())
+{
+    configuration->update(object_storage, context, true, false);
+    auto log = getLogger("IcebergMetadata");
+    auto [last_version, metadata_path, compression_method]
+        = getLatestOrExplicitMetadataFileAndVersion(object_storage, configuration_, nullptr, context_, log.get());
+
+    filename_generator.setVersion(last_version + 1);
+    metadata = getMetadataJSONObject(metadata_path, object_storage, configuration, nullptr, context, log, compression_method);
+
+    partition_spec_id = metadata->getValue<Int64>("default-spec-id");
+    auto partitions_specs = metadata->getArray("partition-specs");
+
+    auto current_schema_id = metadata->getValue<Int64>(Iceberg::f_current_schema_id);
+    Poco::JSON::Object::Ptr current_schema;
+    auto schemas = metadata->getArray(Iceberg::f_schemas);
+    for (size_t i = 0; i < schemas->size(); ++i)
+    {
+        if (schemas->getObject(static_cast<UInt32>(i))->getValue<Int32>(Iceberg::f_schema_id) == current_schema_id)
+        {
+            current_schema = schemas->getObject(static_cast<UInt32>(i));
+        }
+    }
+    for (size_t i = 0; i < partitions_specs->size(); ++i)
+    {
+        auto current_partition_spec = partitions_specs->getObject(static_cast<UInt32>(i));
+        if (current_partition_spec->getValue<Int64>("spec-id") == partition_spec_id)
+        {
+            partititon_spec = current_partition_spec;
+            partitioner = ChunkPartitioner(current_partition_spec->getArray("fields"), current_schema, context_, sample_block_);
+            break;
+        }
+    }
+}
+
+void IcebergStorageSink::consume(Chunk & chunk)
+{
+    if (isCancelled())
+        return;
+    total_rows += chunk.getNumRows();
+    total_chunks_size += chunk.bytes();
+
+    std::unordered_map<ChunkPartitioner::PartitionKey, Chunk, ChunkPartitioner::PartitionKeyHasher> partition_result;
+    if (partitioner)
+    {
+        partition_result = partitioner->participateChunk(chunk);
+    }
+    else
+    {
+        partition_result[{}] = chunk.clone();
+    }
+
+    for (const auto & [partition_key, part_chunk] : partition_result)
+    {
+        if (!data_filenames.contains(partition_key))
+        {
+            auto data_filename = filename_generator.generateDataFileName();
+            data_filenames[partition_key] = data_filename;
+
+            const auto & settings = context->getSettingsRef();
+            const auto chosen_compression_method = chooseCompressionMethod(data_filename, configuration->compression_method);
+
+            auto buffer = object_storage->writeObject(
+                StoredObject(data_filename), WriteMode::Rewrite, std::nullopt, DBMS_DEFAULT_BUFFER_SIZE, context->getWriteSettings());
+
+            write_buffers[partition_key] = wrapWriteBufferWithCompressionMethod(
```

Comment:
Do we really need this? Several times I saw Iceberg data files like `.parquet.zstd`, but does it really make any sense, since parquet support internal compression? Maybe `zstd` in the name mean that the parquet file internally compressed with `zstd`?